src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs

```diff
@@ -298,11 +319,222 @@ void writeEnumElement(string kind, int value, string operationName, string? extr
+        public virtual TResult DefaultVisit(IOperation operation, TArgument argument) => default(TResult);
+        internal virtual TResult VisitNoneOperation(IOperation operation, TArgument argument) => default(TResult);");
+            Indent();
+
+            foreach (var type in types)
+            {
+                if (type.SkipInVisitor) continue;
+
+                WriteObsoleteIfNecessary(type.Obsolete);
+                var accessibility = type.IsInternal ? "internal" : "public";
+                WriteLine($"{accessibility} virtual TResult {GetVisitorName(type)}({type.Name} operation, TArgument argument) => DefaultVisit(operation, argument);");
+            }
+
+            Unbrace();
+            WriteLine("#endregion");
         }
+
+        private void WriteObsoleteIfNecessary(ObsoleteTag? tag)
+        {
+            if (tag is object)
+            {
+                WriteLine($"[Obsolete({tag.Message}, error: {tag.ErrorText})]");
+            }
+        }
+
+        private string GetVisitorName(Node type)
+        {
+            Debug.Assert(!type.SkipInVisitor, $"{type.Name} is marked as skipped in visitor, cannot generate classes.");
+            return type.VisitorName ?? $"Visit{GetSubName(type.Name)}";
+        }
+
+        private List<Property> GetAllProperties(AbstractNode node)
+        {
+            var properties = node.Properties.ToList();
+
+            AbstractNode? @base = node;
+            while (true)
+            {
+                string baseName = @base.Base;
+                @base = _typeMap[baseName];
+                if (@base is null) break;
+                properties.AddRange(@base.Properties);
+            }
+
+            return properties;
+        }
+
+        private string GetSubName(string operationName) => operationName[1..^9];
+
+        private string GetKind(AbstractNode node)
+        {
+            while (node.OperationKind?.Include == false)
+            {
+                node = _typeMap[node.Base] ??
+                    throw new InvalidOperationException($"{node.Name} is not being included in OperationKind, but has no base type!");
+            }
+
+            if (node.OperationKind?.Entries.Count > 0)
+            {
+                return node.OperationKind.Entries.Where(e => e.EditorBrowsable != false).Single().Name;
+            }
+
+            return GetSubName(node.Name);
+        }
+
+        private bool IsIOperationType(string typeName) => _typeMap.ContainsKey(typeName) ||
+                                                          (IsImmutableArray(typeName, out var innerType) && IsIOperationType(innerType));
+
+        private bool IsImmutableArray(string typeName, [NotNullWhen(true)] out string? arrayType)
+        {
+            if (typeName.StartsWith("ImmutableArray<", StringComparison.Ordinal))
+            {
+                arrayType = typeName[15..^1];
```

Comment:
consider extracting out a local `const string immutableArrayPrefix` and using the `.Length` in this slice